agent/agent.go

```diff
@@ -0,0 +1,635 @@
+		if err := os.Remove(o.executablePath); err != nil {
+			return nil, grpc.Errorf(codes.Aborted, "unable to remove existing executable [%s]: %v", o.executablePath, err)
+		}
+		o.executablePath = ""
+	}
+
+	if o.configPath != "" {
+		if err := os.Remove(o.configPath); err != nil {
+			return nil, grpc.Errorf(codes.Aborted, "unable to remove existing config [%s]: %v", o.configPath, err)
+		}
+		o.configPath = ""
+	}
+
+	if err := o.opts.InitHostResourcesFn()(); err != nil {
+		return nil, grpc.Errorf(codes.Internal, "unable to initialize host resources: %v", err)
+	}
+
+	// setup new heartbeating
+	if request.HeartbeatEnabled {
+		beater, err := newHeartbeater(o, request.OperatorUuid, request.HeartbeatEndpoint, o.opts.InstrumentOptions())
+		if err != nil {
+			return nil, grpc.Errorf(codes.Aborted, "unable to start heartbeating process: %v", err)
+		}
+		o.heartbeater = beater
+		o.heartbeater.start(time.Second * time.Duration(request.HeartbeatFrequencySecs))
+	}
+
+	return &m3em.SetupResponse{}, nil
+}
+
+func (o *opAgent) Transfer(stream m3em.Operator_TransferServer) error {
+	o.logger.Infof("receieved Transfer()")
+	var (
+		checksum     = uint32(0)
+		numChunks    = 0
+		lastChunkIdx = int32(0)
+		fileHandle   *os.File
+		fileType     = m3em.FileType_UNKNOWN
+		err          error
+	)
+
+	defer func() {
+		if fileHandle != nil {
+			fileHandle.Close()
+		}
+	}()
+
+	for {
+		request, streamErr := stream.Recv()
+		if streamErr != nil && streamErr != io.EOF {
+			return streamErr
+		}
+
+		if request == nil {
+			o.markFileDone(fileType, fileHandle.Name())
+			return stream.SendAndClose(&m3em.TransferResponse{
+				FileChecksum:   checksum,
+				NumChunksRecvd: int32(numChunks),
+			})
+		}
+
+		if numChunks == 0 {
+			// first request with any data in it, log it for visibilty
+			o.logger.Infof("file transfer initiated: [ filename = %s, fileType = %s, overwrite = %s ]",
+				request.GetFilename(), request.GetType().String(), request.GetOverwrite())
+
+			fileType = request.GetType()
+			fileHandle, err = o.initFile(fileType, request.GetFilename(), request.GetOverwrite())
+			if err != nil {
+				return err
+			}
+			lastChunkIdx = request.GetChunkIdx() - 1
+		}
+
+		chunkIdx := request.GetChunkIdx()
+		if chunkIdx != int32(1+lastChunkIdx) {
+			return fmt.Errorf("received chunkIdx: %d after %d", chunkIdx, lastChunkIdx)
+		}
+		lastChunkIdx = chunkIdx
+
+		numChunks++
+		bytes := request.GetChunkBytes()
+		checksum = crc32.Update(checksum, crc32.IEEETable, bytes)
+
+		numWritten, err := fileHandle.Write(bytes)
+		if err != nil {
+			return err
+		}
+
+		if numWritten != len(bytes) {
+			return fmt.Errorf("unable to write bytes, expected: %d, observed: %d", len(bytes), numWritten)
+		}
+
+		if streamErr == io.EOF {
+			o.markFileDone(fileType, fileHandle.Name())
+			return stream.SendAndClose(&m3em.TransferResponse{
+				FileChecksum:   checksum,
+				NumChunksRecvd: int32(numChunks),
+			})
+		}
+	}
+}
+
+type opAgentHeartBeater struct {
```

Comment:
I can't think of a useful abstraction that two places could share because of how they function: the agent routine is a go-routine periodically sending a heartbeat msg in the background; the operator heartbeating logic polls for messages received (or not received within a timeout period) and propagates that to an event handler.

That said I'd be all for it if you have a suggestion.